### PR TITLE
Support using l1 mintable native token with non-native staking token

### PR DESCRIPTION
**Pull request by dusannosovic-ethernal**
# Description

This PR includes support for l1-mintable token with non-native (erc20) stake token.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [X] I have assigned this PR to myself
- [X] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: command/genesis/polybft_params.go [hunk 1]
```diff
@@ -606,6 +606,15 @@ func (p *genesisParams) getValidatorAccounts() ([]*validator.GenesisValidator, e
...
 			}
 
+			if p.stakeTokenAddr != contracts.NativeERC20TokenContract {
+				s, exists := p.stakeInfos[addr]
+				if !exists {
+					stake = command.DefaultStake
+				} else {
+					stake = s
+				}
+			}
```
Comment (Stefan-Ethernal) [inline command/genesis/polybft_params.go]: Merge this part with the previous check, because the logic for `stakeInfos` population is the same:

```suggestion
			if p.nativeTokenConfig.IsMintable || p.stakeTokenAddr != contracts.NativeERC20TokenContract {
				s, exists := p.stakeInfos[addr]
				if !exists {
					stake = command.DefaultStake
				} else {
					stake = s
				}
			}
```

File: consensus/polybft/polybft.go [hunk 2]
```diff
@@ -204,6 +204,12 @@ func GenesisPostHookFactory(config *chain.Chain, engineName string) func(txn *st
 
 		bridgeCfg := polyBFTConfig.Bridge
 		if bridgeCfg != nil {
+			if polyBFTConfig.StakeTokenAddr != contracts.NativeERC20TokenContract {
```
Comment (Stefan-Ethernal) [inline consensus/polybft/polybft.go]:
```suggestion
			if !isNativeStakeToken(polyBFTConfig) {
```

File: consensus/polybft/contracts_initializer.go [hunk 3]
```diff
@@ -404,3 +429,8 @@ func callContract(from, to types.Address, input []byte, contractName string, tra
 func isNativeRewardToken(cfg PolyBFTConfig) bool {
 	return cfg.RewardConfig.TokenAddress == contracts.NativeERC20TokenContract
 }
+
+// isNativeStakeToken return true in case a native token is used for staking
+func isNativeStakeToken(cfg PolyBFTConfig) bool {
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contracts_initializer.go]: Export the function and use it in `polybft_params.go` as well. Also I'd consider changing so that it accepts `stakeTokenAddr` (as little as possible)


```suggestion
func IsNativeStakeToken(stakeTokenAddr types.Address) bool {
```

File: consensus/polybft/contracts_initializer.go [hunk 4]
```diff
@@ -48,7 +48,7 @@ func initStakeManager(polyBFTConfig PolyBFTConfig, transition *state.Transition)
 			return fmt.Errorf("NativeERC20.approve params encoding failed: %w", err)
 		}
 
-		err = callContract(validator.Address, contracts.NativeERC20TokenContract, input, "NativeERC20.approve", transition)
+		err = callContract(validator.Address, polyBFTConfig.StakeTokenAddr, input, "NativeERC20.approve", transition)
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contracts_initializer.go]:
```suggestion
		err = callContract(validator.Address, polyBFTConfig.StakeTokenAddr, input, "StakeToken.approve", transition)
```

File: consensus/polybft/contracts_initializer.go [hunk 5]
```diff
@@ -367,6 +367,31 @@ func mintRewardTokensToWallet(polyBFTConfig PolyBFTConfig, transition *state.Tra
 		"RewardToken.mint", transition)
 }
 
+func mintStakeToken(polyBFTConfig PolyBFTConfig, transition *state.Transition) error {
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contracts_initializer.go:371]: Consider adding a comment.

File: command/genesis/predeploy/params.go [hunk 6]
```diff
@@ -75,6 +92,12 @@ func (p *predeployParams) initPredeployAddress() error {
 	return nil
 }
 
+func (p *predeployParams) initDeployerAddress() error {
+	p.deployerAddr = types.StringToAddress(p.deployerAddrRaw)
```
Comment (goran-ethernal) [inline command/genesis/predeploy/params.go]: Why not use the `types.IsValidAddr`?

File: consensus/polybft/contracts_initializer.go [hunk 7]
```diff
@@ -404,3 +430,8 @@ func callContract(from, to types.Address, input []byte, contractName string, tra
 func isNativeRewardToken(cfg PolyBFTConfig) bool {
 	return cfg.RewardConfig.TokenAddress == contracts.NativeERC20TokenContract
 }
+
+// isNativeStakeToken return true in case a native token is used for staking
+func IsNativeStakeToken(stakeTokenAddr types.Address) bool {
```
Comment (goran-ethernal) [inline consensus/polybft/contracts_initializer.go:435]: No need to export it:

```suggestion
func isNativeStakeToken(stakeTokenAddr types.Address) bool {
```

File: consensus/polybft/polybft.go [hunk 8]
```diff
@@ -202,6 +197,17 @@ func GenesisPostHookFactory(config *chain.Chain, engineName string) func(txn *st
 			return err
 		}
 
+		if !IsNativeStakeToken(polyBFTConfig.StakeTokenAddr) && polyBFTConfig.Bridge != nil {
```
Comment (goran-ethernal) [inline consensus/polybft/polybft.go]: Why do you need to check if Bridge is enabled?

File: e2e-polybft/e2e/bridge_test.go [hunk 9]
```diff
@@ -1442,3 +1442,112 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+	)
+
+	var (
+		stakeAmount = ethgo.Ether(500)
+	)
+
+	minter, err := wallet.GenerateKey()
+	require.NoError(t, err)
+
+	receiverKeys := make([]string, transfersCount)
```
Comment (goran-ethernal) [inline e2e-polybft/e2e/bridge_test.go]: Why do you need these transfers, receivers, amounts? You are not using this anywhere/

File: e2e-polybft/framework/test-cluster.go [hunk 10]
```diff
@@ -456,6 +458,12 @@ func WithRewardWallet(rewardWallet string) ClusterOption {
 	}
 }
 
+func WithPredeploy() ClusterOption {
+	return func(h *TestClusterConfig) {
+		h.PredeployNonNative = true
```
Comment (goran-ethernal) [inline e2e-polybft/framework/test-cluster.go]: I am not a fan of this approach. What I would do is rename this `PredeployNonNative` to `PredeployContract` which would be a string `<address:artifactName>`, or a small struct like:

```go
    type predeployInfo struct {
         address types.Address
         artifactName string
    }

```

That way, we can use this to pre-deploy whatever we want in tests, and not just some fixed address and fixed artifact.

File: helper/predeployment/predeployment.go [hunk 11]
```diff
@@ -172,24 +174,38 @@ func getPredeployAccount(address types.Address, input []byte, chainID int64) (*c
...
-	if err != nil {
-		return nil, fmt.Errorf("unable to create contract ABI, %w", err)
-	}
+	var constructorInfo *abi.Method
 
-	finalBytecode := artifact.Bytecode
-	constructorInfo := contractABI.Constructor
+	var contractABI *abi.ABI
+
+	if filepath == "" {
```
Comment (goran-ethernal) [inline helper/predeployment/predeployment.go]:
```suggestion
	if conArtifact != nil {
```

**Review by goran-ethernal — CHANGES_REQUESTED** [review]

File: command/genesis/predeploy/params.go (hunk 6 above)
Comment (dusannosovic-ethernal) [reply command/genesis/predeploy/params.go]: Because i didn't rebase my branch on develop, which contains changes for `isValidAddr`, i will do that.

File: command/genesis/predeploy/params.go [hunk 12]
```diff
@@ -8,16 +8,20 @@ import (
...
 const (
-	chainFlag            = "chain"
-	predeployAddressFlag = "predeploy-address"
-	artifactsPathFlag    = "artifacts-path"
-	constructorArgsPath  = "constructor-args"
+	chainFlag               = "chain"
+	predeployAddressFlag    = "predeploy-address"
+	artifactsNameFlag       = "artifacts-name"
+	artifactsPathFlag       = "artifacts-path"
+	constructorArgsPathFlag = "constructor-args"
```
Comment (Stefan-Ethernal) [inline command/genesis/predeploy/params.go]:
```suggestion
	constructorArgsFlag = "constructor-args"
```

File: command/genesis/predeploy/params.go [hunk 13]
```diff
@@ -32,28 +36,37 @@ var (
...
-	genesisConfig *chain.Chain
+	genesisConfig    *chain.Chain
+	contractArtifact *artifact.Artifact
 }
 
 func (p *predeployParams) getRequiredFlags() []string {
 	return []string{
 		predeployAddressFlag,
-		artifactsPathFlag,
+		artifactsNameFlag,
```
Comment (Stefan-Ethernal) [inline command/genesis/predeploy/params.go]: I'd still require artifactsPathFlag as it was before. Providing artifact via the name is optional convenience.

```suggestion
		artifactsPathFlag,
```

File: command/genesis/predeploy/params.go (hunk 6 above)
Comment (Stefan-Ethernal) [reply command/genesis/predeploy/params.go]: Reminder to change this. As the mater of fact we dont't need this function anymore when we change to use `types.IsValidAddr`.

File: e2e-polybft/framework/test-cluster.go (hunk 10 above)
Comment (Stefan-Ethernal) [reply e2e-polybft/framework/test-cluster.go]: Reminder to address this comment @dusannosovic-ethernal

File: consensus/polybft/contracts_initializer.go [hunk 14]
```diff
@@ -48,7 +48,7 @@ func initStakeManager(polyBFTConfig PolyBFTConfig, transition *state.Transition)
 			return fmt.Errorf("NativeERC20.approve params encoding failed: %w", err)
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contracts_initializer.go]:
```suggestion
			return fmt.Errorf("Staking ERC20.approve params encoding failed: %w", err)
```

File: consensus/polybft/contracts_initializer.go [hunk 15]
```diff
@@ -48,7 +48,7 @@ func initStakeManager(polyBFTConfig PolyBFTConfig, transition *state.Transition)
...
 		}
 
-		err = callContract(validator.Address, contracts.NativeERC20TokenContract, input, "NativeERC20.approve", transition)
+		err = callContract(validator.Address, polyBFTConfig.StakeTokenAddr, input, "StakeToken.approve", transition)
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contracts_initializer.go]:
```suggestion
		err = callContract(validator.Address, polyBFTConfig.StakeTokenAddr, input, "Staking ERC20.approve", transition)
```

File: consensus/polybft/contracts_initializer.go [hunk 16]
```diff
@@ -367,6 +367,32 @@ func mintRewardTokensToWallet(polyBFTConfig PolyBFTConfig, transition *state.Tra
...
+
+	for _, validator := range polyBFTConfig.InitialValidatorSet {
+		mintFn := contractsapi.MintRootERC20Fn{
+			To:     validator.Address,
+			Amount: validator.Stake,
+		}
+
+		input, err := mintFn.EncodeAbi()
+		if err != nil {
+			return fmt.Errorf("StakeToken.mint params encoding failed: %w", err)
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contracts_initializer.go]:
```suggestion
			return fmt.Errorf("Staking ERC20.mint params encoding failed: %w", err)
```

File: consensus/polybft/contracts_initializer.go [hunk 17]
```diff
@@ -367,6 +367,32 @@ func mintRewardTokensToWallet(polyBFTConfig PolyBFTConfig, transition *state.Tra
...
+		}
+
+		if err := callContract(polyBFTConfig.BladeAdmin, polyBFTConfig.StakeTokenAddr,
+			input, "StakeToken.mint", transition); err != nil {
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contracts_initializer.go]:
```suggestion
			input, "Staking ERC20.mint", transition); err != nil {
```

File: consensus/polybft/contractsapi/init.go [hunk 18]
```diff
@@ -14,306 +15,361 @@ const (
...
-	CustomSupernetManager           *artifact.Artifact
-	StakeManager                    *artifact.Artifact
-	EpochManager                    *artifact.Artifact
-	RootERC721                      *artifact.Artifact
-	RootERC1155                     *artifact.Artifact
-	EIP1559Burn                     *artifact.Artifact
-	BladeManager                    *artifact.Artifact
-	GenesisProxy                    *artifact.Artifact
-	TransparentUpgradeableProxy     *artifact.Artifact
+	CheckpointManager               *contracts.Artifact
```
Comment (Stefan-Ethernal) [inline consensus/polybft/contractsapi/init.go:18]: We should get rid off all these `contracts.Artifact` fields and read them from the map. However, that can be a separate PR.

File: command/genesis/predeploy/params.go [hunk 19]
```diff
@@ -26,34 +30,48 @@ var (
...
 
-func (p *predeployParams) initRawParams() error {
+func (p *predeployParams) initRawParams() (err error) {
+	if p.artifactsName == "" && p.artifactsPath == "" {
+		return errArtifactPathAndNameMissing
+	}
+
 	if err := p.initPredeployAddress(); err != nil {
 		return err
 	}
```
Comment (Stefan-Ethernal) [inline command/genesis/predeploy/params.go]:
```suggestion
	var err error  
    
    p.address, err = types.IsValidAddr()  
    if err != nil {
        return errInvalidPredeployAddress
```

Also `initPredeployAddress` can be removed in that case.

File: command/genesis/predeploy/genesis_predeploy.go [hunk 20]
```diff
@@ -38,6 +38,13 @@ func setFlags(cmd *cobra.Command) {
 		fmt.Sprintf("the address to predeploy to. Must be >= %s", predeployAddressMin.String()),
 	)
 
+	cmd.Flags().StringVar(
+		&params.artifactsName,
+		artifactsNameFlag,
+		"",
+		"the path to the contract artifacts JSON",
```
Comment (Stefan-Ethernal) [inline command/genesis/predeploy/genesis_predeploy.go]:
```suggestion
		"the built-in contract artifact name",
```

File: e2e-polybft/e2e/bridge_test.go [hunk 21]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
 		}
 	})
 }
+
+func TestE2E_Bridge_NonNative(t *testing.T) {
```
Comment (Stefan-Ethernal) [inline e2e-polybft/e2e/bridge_test.go]:
```suggestion
func TestE2E_Bridge_NonNativeStakingToken(t *testing.T) {
```

File: e2e-polybft/e2e/bridge_test.go [hunk 22]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+	validatorAcc, err := validatorHelper.GetAccountFromDir(firstValidator.DataDir())
+	require.NoError(t, err)
+
+	relayer, err := txrelayer.NewTxRelayer(txrelayer.WithIPAddress(firstValidator.JSONRPCAddr()))
+	require.NoError(t, err)
+
+	mintFn := &contractsapi.MintRootERC20Fn{
+		To:     validatorAcc.Address(),
+		Amount: big.NewInt(100000),
+	}
```
Comment (Stefan-Ethernal) [inline e2e-polybft/e2e/bridge_test.go:1496]:
```suggestion
		Amount: ethgo.Gwei(10),
```

File: e2e-polybft/e2e/bridge_test.go [hunk 23]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+	polybftCfg, err := polybft.LoadPolyBFTConfig(path.Join(cluster.Config.TmpDir, chainConfigFileName))
+	require.NoError(t, err)
+
+	firstValidatorInfo, err := validatorHelper.GetValidatorInfo(validatorAcc.Ecdsa.Address(), relayer)
+	require.NoError(t, err)
+	require.True(t, firstValidatorInfo.Stake.Cmp(stakeAmount) == 0)
+
+	require.NoError(t, cluster.WaitForBlock(epochSize*1, 1*time.Minute))
+
+	require.NoError(t, firstValidator.Stake(polybftCfg, big.NewInt(1000), contracts.ERC20Contract))
```
Comment (Stefan-Ethernal) [inline e2e-polybft/e2e/bridge_test.go]:
```suggestion
	require.NoError(t, firstValidator.Stake(polybftCfg, ethgo.Gwei(1), contracts.ERC20Contract))
```

also extract stake amount that you are staking to a variable. Not sure can we reuse `stakeAmount`?

File: e2e-polybft/e2e/bridge_test.go [hunk 24]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+
+	require.NoError(t, cluster.WaitForBlock(epochSize*3, 90*time.Second))
+
+	updatedStakeAmount := big.NewInt(0)
+
+	firstValidatorInfo, err = validatorHelper.GetValidatorInfo(validatorAcc.Ecdsa.Address(), relayer)
+	require.NoError(t, err)
+	require.True(t, firstValidatorInfo.Stake.Cmp(updatedStakeAmount.Add(stakeAmount, big.NewInt(1000))) == 0)
+
+	require.NoError(t, firstValidator.Unstake(big.NewInt(1000)))
```
Comment (Stefan-Ethernal) [inline e2e-polybft/e2e/bridge_test.go]: Seems like you are unstaking entire stake, so reuse stake amount variable that you are about to introduce.

File: e2e-polybft/e2e/bridge_test.go [hunk 25]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+
+	require.NoError(t, firstValidator.Stake(polybftCfg, big.NewInt(1000), contracts.ERC20Contract))
+
+	require.NoError(t, cluster.WaitForBlock(epochSize*3, 90*time.Second))
+
+	updatedStakeAmount := big.NewInt(0)
+
+	firstValidatorInfo, err = validatorHelper.GetValidatorInfo(validatorAcc.Ecdsa.Address(), relayer)
+	require.NoError(t, err)
+	require.True(t, firstValidatorInfo.Stake.Cmp(updatedStakeAmount.Add(stakeAmount, big.NewInt(1000))) == 0)
```
Comment (Stefan-Ethernal) [inline e2e-polybft/e2e/bridge_test.go]: Instead of `big.NewInt(1000)`, reuse variable that you are about to introduce. That way, code is more readable.

File: e2e-polybft/e2e/bridge_test.go [hunk 26]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+	nonNativeErc20 := ethgo.Address(contracts.ERC20Contract)
+
+	receipt, err := relayer.SendTransaction(
+		&ethgo.Transaction{
+			To:    &nonNativeErc20,
+			Input: mintInput,
+			Type:  ethgo.TransactionDynamicFee,
+		}, minter)
+	require.NoError(t, err)
+	require.Equal(t, uint64(types.ReceiptSuccess), receipt.Status)
```
Comment (Stefan-Ethernal) [inline e2e-polybft/e2e/bridge_test.go:1503]: I don't understand this part, seems like you are minting to the token itself (instead to some other account, such as validator or externally owned account)? If that is the case, it seems like unnecesary step.

```suggestion
```

File: e2e-polybft/e2e/bridge_test.go [hunk 27]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+		To:     validatorAcc.Address(),
+		Amount: big.NewInt(100000),
+	}
+
+	mintInput, err := mintFn.EncodeAbi()
+	require.NoError(t, err)
+
+	nonNativeErc20 := ethgo.Address(contracts.ERC20Contract)
+
+	receipt, err := relayer.SendTransaction(
```
Comment (goran-ethernal) [inline e2e-polybft/e2e/bridge_test.go:1503]: You don't need to manually create a `mint` transaction. On `test-server` you have `MintNativeERC20Token` function. You can modify it to accept a `erc20` token address, and call it in your test with your stake token address.

File: e2e-polybft/e2e/bridge_test.go [hunk 28]
```diff
@@ -1442,3 +1442,90 @@ func TestE2E_Bridge_NonMintableERC20Token_WithPremine(t *testing.T) {
...
+
+	mintFn := &contractsapi.MintRootERC20Fn{
+		To:     validatorAcc.Address(),
+		Amount: big.NewInt(100000),
+	}
+
+	mintInput, err := mintFn.EncodeAbi()
+	require.NoError(t, err)
+
+	nonNativeErc20 := ethgo.Address(contracts.ERC20Contract)
```
Comment (Stefan-Ethernal) [inline e2e-polybft/e2e/bridge_test.go]: You shouldn't hardcode the address to ERC20Contract, but use the one in the `PolyBFTConfig`.

Also, let's remove `contracts.ERC20Contract` altogether, and as @goran-ethernal explained in one of the comments, once we can provide the address and contract that should be pre-deployed, we can use that address from the variable here.

**Review by Stefan-Ethernal — DISMISSED** [review]

File: consensus/polybft/polybft.go [hunk 29]
```diff
@@ -202,6 +197,17 @@ func GenesisPostHookFactory(config *chain.Chain, engineName string) func(txn *st
 			return err
 		}
 
+		if !IsNativeStakeToken(polyBFTConfig.StakeTokenAddr) {
```
Comment (goran-ethernal) [inline consensus/polybft/polybft.go]: You don't need this condition. `mintStakeToken` has this check.

File: e2e-polybft/framework/test-server.go [hunk 30]
```diff
@@ -285,8 +288,8 @@ func (t *TestServer) WhitelistValidators(addresses []string) error {
 	return runCommand(t.clusterConfig.Binary, args, t.clusterConfig.GetStdout("validator"))
 }
 
-// MintNativeERC20Token mints given amounts of native erc20 token on blade to given addresses
-func (t *TestServer) MintNativeERC20Token(addresses []string, amounts []*big.Int) error {
+// MintERC20Token mints given amounts of native erc20 token on blade to given addresses
+func (t *TestServer) MintERC20Token(addresses []string, amounts []*big.Int, stakeToken types.Address) error {
```
Comment (goran-ethernal) [inline e2e-polybft/framework/test-server.go]:
```suggestion
func (t *TestServer) MintERC20Token(addresses []string, amounts []*big.Int, erc20Token types.Address) error {
```

**Review by goran-ethernal — DISMISSED** [review]
There are some small comments left, but in general looks good.

**Review by Stefan-Ethernal — APPROVED** [review]
